File: pdoo/java/Deepspace/EnemyStarShip.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

/**
 * @brief Clase nave enemiga
 */
class EnemyStarShip implements Copyable <EnemyStarShip>, SpaceFighter {
    private Damage damage;
    private Loot loot;
    private String name;
    private float ammoPower;
    private float shieldPower;
    
    EnemyStarShip(String n, float a, float s,Loot l,Damage d) {
        damage = d;
        loot = l;
        name = n;
        ammoPower = a;
        shieldPower = s;
    }
    
    EnemyStarShip(EnemyStarShip e) {
        damage = e.damage;
        loot = e.loot;
        name = e.name;
        ammoPower = e.ammoPower;
        shieldPower = e.shieldPower;
    }
    
    EnemyToUI getUIversion() { return new EnemyToUI(this); }
    
    @Override
    public float fire() { return ammoPower; }
    
    public float getAmmoPower() { return ammoPower; }
    
    public Damage getDamage() { return damage;  }
    
    public Loot getLoot() { return loot; }
    
    public String getName() { return name; }
    
    public float getShieldPower() { return shieldPower; }
    
    @Override
    public float protection() { return shieldPower; }
    
    @Override
    public ShotResult recieveShot(float shot) {
        if (shieldPower<shot)
            return ShotResult.DONOTRESIST;
        else
            return ShotResult.RESIST;
    }
    
    @Override
    public String toString() {
        String aux = "Nombre del enemigo: "+name;
        aux += "\nPoder de disparo: "+ammoPower;
        aux += "\nPoder de escudo: "+shieldPower;
        aux += "\nDaño de la nave:\n"+damage.toString();
        aux += "\nBotín de la nave:\n"+loot.toString();
        return aux;
    }

    @Override
    public EnemyStarShip copy() {
       return new EnemyStarShip(this);
    }
}

File: pdoo/ruby/enemystarship.rb
#encoding:utf-8

require "./lib/EnemyToUI.rb"
require_relative "./shotresult.rb"
require_relative "./damage.rb"
require_relative "./loot.rb"

module Deepspace
  # Nave enemiga
  class EnemyStarShip

    attr_reader :name, :ammoPower, :shieldPower, :loot, :damage

    def initialize(n,a,s,l,d)
      @name = n
      @ammoPower = a
      @shieldPower = s
      @loot = l
      @damage = d
    end

    # Constructor por copia
    def self.newCopy(e)
      EnemyStarShip.new(e.name,e.ammoPower,e.shieldPower,e.loot,e.damage)
    end

    # Análogo al consultor del atributo 'ammoPower'
    def fire
      @ammoPower
    end

    # Análogo al consultor del atributo 'shieldPower'
    def protection
      @shieldPower
    end
    
    # Determina si la nave aguanta un disparo
    def recieveShot(shot)
      if @shieldPower<shot
        return ShotResult::DONOTRESIST
      else
        return ShotResult::RESIST
      end
    end

    def getUIversion
      EnemyToUI.new(self)
    end

    def to_s
      getUIversion.to_s
    end
  end
end
